perceval/backends/mediawiki.py

```diff
@@ -0,0 +1,475 @@
+        for review in reviews:
+            ts = str_to_datetime(review['timestamp'])
+            ts = ts.replace(tzinfo=dateutil.tz.tzutc())
+            if ts.timestamp() > max_ts:
+                max_ts = ts.timestamp()
+        return max_ts
+
+    @metadata
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the pages from the backend url.
+
+        The method retrieves, from a MediaWiki url, the
+        wiki pages.
+
+        :returns: a generator of pages
+        """
+
+        self._purge_cache_queue()
+
+        if from_date == DEFAULT_DATETIME:
+            from_date = None
+        else:
+            from_date = datetime_to_utc(from_date).replace(tzinfo=None)
+
+        mediawiki_version = self.client.get_version()
+        logging.info("MediaWiki version: %s", mediawiki_version)
+
+        if mediawiki_version[0] >= 1 and mediawiki_version[1] >= 27:
+            fetcher = self.__fetch_1_27(from_date)
+        else:
+            fetcher = self.__fetch_pre1_27(from_date)
+
+        for page_reviews in fetcher:
+            yield page_reviews
+
+
+    def __fetch_1_27(self, from_date=None):
+        """Fetch the pages from the backend url for MediaWiki >=1.27
+
+        The method retrieves, from a MediaWiki url, the
+        wiki pages.
+
+        :returns: a generator of pages
+        """
+
+        logger.info("Looking for pages at url '%s'", self.url)
+
+        self._purge_cache_queue()
+        npages = 0  # number of pages processed
+
+        # The pages are organized in content namespaces
+        raw_namespaces = self.client.get_namespaces()
+        namespaces = json.loads(raw_namespaces)["query"]["namespaces"]
+        namespaces_contents = [ns for ns in namespaces if 'content' in namespaces[ns].keys()]
+        self._push_cache_queue(raw_namespaces)
+        self._flush_cache_queue()
+
+        arvcontinue = '' # pagination for getting revisions and their pages
+        while arvcontinue is not None:
+            raw_pages = self.client.get_pages_from_allrevisions(namespaces_contents, from_date, arvcontinue)
+            self._push_cache_queue(raw_pages)
+            self._flush_cache_queue()
+            data_json = json.loads(raw_pages)
+            if 'continue' in data_json:
+                arvcontinue = data_json['continue']['arvcontinue']
+            else:
+                arvcontinue = None
+            pages_json = data_json['query']['allrevisions']
+            for page in pages_json:
+                urltitle = urllib.parse.urlencode({'titles':page['title']})
+                revisions_raw = self.client.get_revisions(urltitle)
+                self._push_cache_queue(revisions_raw)
+                self._flush_cache_queue()
+                page_reviews = self.__build_page_reviews(page, json.loads(revisions_raw))
+                yield page_reviews
+                npages += 1
+
+        logger.info("Total number of pages: %i", npages)
+
+
+
+    def __fetch_pre1_27(self, from_date=None):
```

Comment:
This method is too long. I think you can split it in several methods for a better understanding.